### ISSUE TITLE
Using HTTP Proxy: Retrieved Host is IP Address instead of Domain in HTTPS Tracker URL

### ISSUE THREAD

**Issue by bearx3f** [arvidn/libtorrent#7651]
**Please provide the following information**

libtorrent version (or branch): 2.0.10

platform/architecture: qbittorrent-static-nox x86_64

compiler and compiler version: GCC 13.2.0 -- x86_64-linux-musl

I'm using qbittorrent and having http proxy settings 

![image](https://github.com/arvidn/libtorrent/assets/9717561/699b2bfb-3448-47cb-9e12-1a221abe09cf)

And when announce to HTTPS Tracker libtorrent not sent Host name on CONNECT request

**Comment by arvidn** [comment]
correct. the `CONNECT` command optionally includes `Proxy-Authorization:`, but not the hostname of the destination.

The `Host` header is sent inside the SSL tunnel, to the destination server.

Is there a problem? Does the proxy need to know the hostname (as opposed to just the IP address) of the destination server?

**Comment by bearx3f** [comment]
there's a problem to sign the ssl connection in mitm mode proxy

**Comment by arvidn** [comment]
I'm intrigued. What's the feature or utility that's enabled by the proxy man-in-the-middling the connection?

You would also have to install their SSL cert for libtorrent to pick up as well. Is that a certification signed for all hostnames?

**Comment by bearx3f** [comment]
> What's the feature or utility that's enabled by the proxy man-in-the-middling the connection?

I'm tracking the download, upload and event of the torrent

> Is that a certification signed for all hostnames?

No, just sign for each hostname of each request

**Comment by arvidn** [comment]
@bearx3f could you try this patch out?
https://github.com/arvidn/libtorrent/pull/7710

**Comment by bearx3f** [comment]
Thank you, for your hard work I'll try

**Comment by bearx3f** [comment]
I'm sorry I cannot test your code but after review your change I think may you change code at http_connection.cpp@591

```cpp
	if (m_proxy.send_host_in_connect)
	{
		if (boost::get<http_stream>(&*m_sock))
		{
			boost::get<http_stream>(*m_sock).set_host(m_hostname);
		}
#if TORRENT_USE_SSL
		else if (m_ssl && boost::get<ssl_stream<http_stream>>(&*m_sock))
		{
			boost::get<ssl_stream<http_stream>>(*m_sock).next_layer().set_host(m_hostname);
		}
#endif
	}
```

**Comment by bearx3f** [comment]
I've made an edit #8021, and after testing, it seems we don't need to send the Host header at all. The line `CONNECT example.com:443 HTTP/1.0` is sufficient to tell the proxy server what `domain:port` it should connect to.

If that's the case, I found a suitable setting in setting_pack called proxy_hostnames, which means "Perform hostname lookup via proxy." The comment for this setting says, "if true, hostname lookups are done via the configured proxy (if any). This is only supported by SOCKS5 and HTTP."

However, after reviewing the code, I found it only supports sock5 and sock5_pw, for example, at http_connection.cpp:371. I'm not sure which direction I should take for the fix:

- Keep the `send_host_in_connect` setting, which is used for http and http_pw proxy types.

- Modify the `proxy_hostnames` setting to also support http and http_pw.

**Comment by bearx3f** [comment]
After studying the RFCs about the CONNECT request, I found that to make the request fully compliant, it must include the Host header, since some proxy servers will reject a CONNECT request if it is incomplete.
